python/cudf/cudf/core/column/column.py

```diff
@@ -365,26 +389,64 @@ def nullmask(self) -> Buffer:
             raise ValueError("Column has no null mask")
         return self.mask_array_view
 
+    @property
+    def _nullmask(self) -> Buffer:
+        """The gpu buffer for the null-mask"""
+        if not self.nullable:
+            raise ValueError("Column has no null mask")
+        return self._mask_array_view
+
+    def force_deep_copy(self: T) -> T:
+        """
+        A method to create deep copy irrespective of whether
+        `copy-on-write` is enable.
+        """
+        result = libcudf.copying.copy_column(self)
+        return cast(T, result._with_type_metadata(self.dtype))
+
     def copy(self: T, deep: bool = True) -> T:
         """Columns are immutable, so a deep copy produces a copy of the
         underlying data and mask and a shallow copy creates a new column and
         copies the references of the data and mask.
+
+        Note : Only Fixed width columns are mutable i.e., which support
+        creation of a `mutable_view`.
         """
         if deep:
-            result = libcudf.copying.copy_column(self)
-            return cast(T, result._with_type_metadata(self.dtype))
+            return self.force_deep_copy()
         else:
-            return cast(
-                T,
-                build_column(
-                    self.base_data,
-                    self.dtype,
-                    mask=self.base_mask,
-                    size=self.size,
-                    offset=self.offset,
-                    children=self.base_children,
-                ),
-            )
+            if cudf.get_option("copy_on_write"):
+
+                copied_col = cast(
+                    T,
+                    build_column(
+                        data=self.base_data
+                        if self.base_data is None
+                        else self.base_data.copy(deep=deep),
+                        dtype=self.dtype,
+                        mask=self.base_mask
+                        if self.base_mask is None
+                        else self.base_mask.copy(deep=deep),
+                        size=self.size,
+                        offset=self.offset,
+                        children=tuple(
+                            col.copy(deep=False) for col in self.base_children
+                        ),
+                    ),
+                )
+                return copied_col
+            else:
+                return cast(
+                    T,
+                    build_column(
+                        self.base_data,
+                        self.dtype,
+                        mask=self.base_mask,
+                        size=self.size,
+                        offset=self.offset,
+                        children=self.base_children,
+                    ),
+                )
```

Comment:
I think we can simplify this whole function by moving the conditional into the logic for the children parameter. I can't comment since only parts of the section are modified, but basically I think this definition should be equivalent:
```
        if deep:
            return self.force_deep_copy()
        else:
            return cast(
                T,
                build_column(
                    data=self.base_data
                    if self.base_data is None
                    else self.base_data.copy(deep=False),
                    dtype=self.dtype,
                    mask=self.base_mask
                    if self.base_mask is None
                    else self.base_mask.copy(deep=False),
                    size=self.size,
                    offset=self.offset,
                    children=tuple(
                        col.copy(deep=False) for col in self.base_children
                    ) if cudf.get_option("copy_on_write"): else self.base_children,
                ),
            )
```
I think the new impl of `Buffer.copy` should allow calling `copy` with no real extra overhead in the non-COW case and give the right behavior, right?

Reply:
Yup, that's correct! Simplified the implementation.